include/bout/array.hxx

```diff
@@ -120,8 +121,10 @@ public:
 
   /*!
    * Resize the array to \p new_size
+   *
+   * Note that this invalidates the existing data!
    */
-  void resize(int new_size) {
+  void resize(size_type new_size) {
     release(ptr);
     ptr = get(new_size);
```

Comment:
Note if `size()==new_size` then I guess there's a good chance (but no promise - or maybe we can promise this due to only one thread having access to this store at one time) that we actually end up getting the same block back off the store that we just released back to the store. Might be worth making a note of this somewhere -- I could see this leading to confusion/hiding problems potentially. 

Is it worth having some sort of `resizeInPlace` routine that promises to keep the first `min(size(), new_size)` elements of data that would behave a bit more like `resize` on `std::vector`? If we kept track of how big we thought the size of the array was we could make shrinking resizes not potentially do an allocation (we use our idea of size for bounds checking and responding to external size queries -- a shrinking resize just changes this without releasing/getting, a growing resize checks if the new size is bigger than the actual size, if it isn't we can just change our tracked upper bound).

Reply:
For the first point, I'm not sure I see the problem. Do we just need to be more explicit in the docstring that you shouldn't use the values the `Array` immediately after a call to `resize`?

For the second point, I can see the appeal but:
1. this would require a fair amount of changes to the `Array` internals.
2. the major use-case of `Array::resize` is to set the size of a default-initialised member variable. `Array` is not quite a general-purpose container like `std::vector` and I'm not sure we want it to be (similarly for `Matrix/Tensor`).

Reply:
Maybe `resize` is the wrong name, and this should really be `reallocate`?

Reply:
Yes, this makes sense: `resize` does imply that the original contents will be kept.

Reply:
Renamed `resize` -> `reallocate`